services/gateway/gateway-service.mjs

```diff
@@ -210,77 +211,82 @@ class GatewayService {
       }
     }
 
-    // Load the player data.
-    socket.playerData = await this.dir.getPlayerData(playerID);
-    socket.playerID = playerID;
+    // Acquire this lock if you need to move the player.
+    socket.moveLock = new AsyncLock();
+    await socket.moveLock.acquire('move', async () => {
```

Comment:
I think it is overkill to make the whole process locked.
The necessary part, i.e. the code that really deals with moving players, is only `socket.on('avatarSelect', ...`
Note that I suggest a little refactoring in below comment.

Reply:
Added comments on why that is done. (To serialize the connect and disconnect handlers.)